packages/lesswrong/client/apolloClient.ts

```diff
@@ -1,8 +1,110 @@
 import { ApolloClient, NormalizedCacheObject, InMemoryCache, ApolloLink } from '@apollo/client';
 import { createHttpLink, createErrorLink, headerLink } from '../lib/apollo/links';
+import type { FieldFunctionOptions } from '@apollo/client';
+
+interface CacheReference {
+  __ref: string;
+}
+
+// Type for the structure of items within the results array *in the cache*
+// Nested items are likely references after normalization.
+interface CachedUltraFeedResultItem {
+  __typename: 'UltraFeedEntryType';
+  type: 'feedPost' | 'feedCommentThread' | 'feedSpotlight';
+  feedPost?: CacheReference | null; // Primarily expect Reference or null
+  feedCommentThread?: CacheReference | null;
+  feedSpotlight?: CacheReference | null;
+}
+
+// Define the key function logic directly here based on MixedTypeFeed
+const ultraFeedKeyFunc = (result: CachedUltraFeedResultItem, readField: FieldFunctionOptions['readField']) => {
+  let id: string | undefined;
+  let objectRef: CacheReference | null | undefined;
+
+  switch (result.type) {
+    case 'feedPost':
+      objectRef = result.feedPost;
+      break;
+    case 'feedCommentThread':
+      objectRef = result.feedCommentThread;
+      break;
+    case 'feedSpotlight':
+      objectRef = result.feedSpotlight;
+      break;
+  }
+
+  if (objectRef) {
+    id = readField<string>('_id', objectRef);
+  }
+
+  return `${result.type}_${id ?? 'undefined'}`;
+};
 
 export const createApolloClient = (baseUrl = '/'): ApolloClient<NormalizedCacheObject> => {
-  const cache = new InMemoryCache();
+  const cache = new InMemoryCache({
+    typePolicies: {
```

Comment:
Can I get a brief tl;dr?

Reply:
(If changes to the apollo client cache are necessary I'm pretty worried about what's going on elsewhere)

Reply:
I have reverted this since it was to avoid a pre-existing warning